beps(dynamic-discovery): pivot to just plugin listing with static config

aramissennyeydd opened the pull request:
## Hey, I just made a Pull Request!

Pivoting this BEP to remove any of the dynamic discovery scope. 

The new version will use the existing `HostDiscovery` static config (or just the list of installed plugins if not split) and return a list of installed plugins - easy as that 😁. I'm hoping this make alignment quite a bit easier, since the scope is just 2 new services and HTTP APIs to match.

Hoping to use this API for future work to build on.

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

aramissennyeydd committed:
adjustments after working with it

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd committed:
move to implementable

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd committed:
pivoting to just definitive plugin listing

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd committed:
typo

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

benjdlambert commented:
@aramissennyeydd sorry, I've been out on vacation and this one slipped through the cracks a little bit!

Wanna bring in some other @backstage/maintainers too to have a read through :tada:

freben commented on beps/0005-definitive-plugin-listing/README.md:
```diff
@@ -0,0 +1,250 @@
+                        type:
+                          type: string
+                          enum:
+                            - plugin
```
```suggestion
                            - module
```

freben commented on beps/0005-definitive-plugin-listing/README.md:
```diff
@@ -0,0 +1,250 @@
+                        type:
+                          type: string
+                          enum:
+                            - plugin
```
```suggestion
                            - module
```

freben commented on beps/0005-definitive-plugin-listing/README.md:
````diff
@@ -0,0 +1,250 @@
+
+```yaml
+paths:
+  /.backstage/deploymentInfo/features/installed:
````
So this is more for frontend consumption, while the other one is for service consumption - so to speak? As in, if we make a frontend plugin that wants to show an overview of your Backstage ecosystem, would it just pick some random plugin ID of sorts and call this enpoint there? And then that would perhaps use discovery and make a bunch of instance info calls to do its job?

Makes me wonder if the deployment info service shouldn't be a service, but rather a core plugin whose ID you could resolve with discovery and make calls to.

aramissennyeydd replied:
Interesting point, I definitely agree - instance data needs to be a service to tie into startup information, but this could be spun off to a separate plugin for sure.

freben reviewed (commented):
Alright, some thoughts. Pulling in @Rugvip

freben commented on beps/0005-definitive-plugin-listing/README.md:
````diff
@@ -0,0 +1,250 @@
+    };
+```
+
+### New `InstanceMetadataService`
````
One more thought. Making it a service implies the ability to depend on it and possibly even replace it. Does either of those two actions make any sense? Would it be an antipattern to try to enumerate features inside another plugin, while early startup is still in full swing? Essentially - what if this was dialed back to only being the `instanceInfo` HTTP endpoint and no service at all? Would that make sense?

Rugvip replied:
I think it makes sense to split into two services tbh, one that is strictly for colocated plugins and very explicit about that. That one would be provided by the backend initializer, just like the existing plugin metadata service. In addition to that we also add a separate service that is the one with more appealing naming and the one actually intended to be used for discovery. We keep that one replacable though, so that one can plugin in their own way of listing available services across all deployed instances.

aramissennyeydd replied:
@Rugvip Splitting the `InstanceMetadata` service into 2 or using the services described here to create a `BackstageMetadataService` or something that could be used for discovery?

aramissennyeydd committed:
Apply suggestions from code review

Co-authored-by: Fredrik Adelöw <freben@gmail.com>
Signed-off-by: Aramis Sennyey <159921952+aramissennyeydd@users.noreply.github.com>

aramissennyeydd commented on beps/0005-definitive-plugin-listing/README.md:
```diff
@@ -0,0 +1,250 @@
+      pluginId: string;
+    }
+  | {
+      type: 'module';
```
Definitely could include the extension points + service factories, I wasn't familiar with those so didn't include them here, but they likely fit in well.

Loaders could make sense to include in a future iteration, but they may make more sense as a separate API. Similarly, dependencies and their versions _could_ fit into this API or may make sense as a separate API that only includes versioned registrations - I'd imagine something like service factories are likely to be less well versioned.

Rugvip replied:
Let's keep it slim like this for now :>

aramissennyeydd committed:
rename BackstageFeature to BackendFeatureMeta

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd commented on beps/0005-definitive-plugin-listing/README.md:
````diff
@@ -0,0 +1,250 @@
+
+```ts
+const installedFeatures = deploymentMetadataService.getInstalledFeatures();
+// installedFeatures = [{type: 'plugin', pluginId: 'catalog'}, {type: 'plugin', pluginId: 'auth'}, {type: 'plugin', pluginId: 'permission'}, {type: 'plugin', pluginId: 'search'}]
````
As is, this implementation would only return plugins and nothing dynamic since it isn't calling into the individual instance. I'll call that out better here, I was trying to provide a solid static implementation that was extendable once we get more dynamic.

I was trying to scope this to a static implementation for ease of use and less config/routing problems. Happy to go down the route of URL based discovery, but I was thinking that would be a v2 of this BEP and not immediately implemented.

aramissennyeydd replied:
Coming back to this, this is a poor example that matches HostDiscovery but doesn't demonstrate really what I was trying to do here. `moduleId` is difficult to do without the dynamic discovery this BEP had originally. We could update this to URL based, ie initially using the list of URLs configured in the discovery config and working backwards based on their instance metadata endpoints - I would like that to be a v2 of this proposal, it definitely makes sense, but has more moving parts that I think implementing the static version of would help flush out.

aramissennyeydd replied:
Coming back to this again - I think URLs is the right answer here. It solves the issue with modules not showing up in my initial design and ensures that data will be consistent if configs are different across machines. I'll update the doc with that.

aramissennyeydd committed:
apply PR feedback

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd commented:
@freben I think responded to all of your comments, PTAL 🙇

aramissennyeydd committed:
move towards just enumerating instances

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd commented:
reread what I had - tried to simplify it more by just having `DeploymentMetadataService` return a list of instances. Leaving it up to the consumer to decide if/how they want to reach out to each instance. I think that will require a new config property on the existing `discovery` config, `baseUrl` or `metadataUrl` to reach the metadata URL itself as the routing is currently plugin focused.

aramissennyeydd committed:
updates following the SIG

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

Rugvip commented on beps/0005-definitive-plugin-listing/README.md:
```diff
@@ -0,0 +1,240 @@
+know that this has succeeded?
+-->
+
+1. As an integrator, I can now get a list of currently installed features across my deployment.
+1. As an integrator/administrator, I can get a list of currently installed features for a single instance in my deployment.
```
```suggestion
1. As a plugin builder, I can now get a list of currently installed features across my deployment.
1. As a plugin builder, I can get a list of currently installed features for a single instance in my deployment.
```

Right? Really the role that would be doing that. If you want to expose the available features for an integrator/operator you'd need to do that through a plugin.

Rugvip commented on beps/0005-definitive-plugin-listing/README.md:
````diff
@@ -0,0 +1,240 @@
+
+```ts
+interface InstanceMetadataService {
+  getInstalledFeatures: () => BackendFeatureMeta[];
````
Related to the discussion above, the service that returns locally installed features would probably have a method that's called something like `getLocallyInstalledFeatures`, `getColocatedFeatures`, or `getFeaturesInstalledOnTheSameInstance`, something very explicit x)

aramissennyeydd replied:
updated in [05d1f2b](https://github.com/backstage/backstage/pull/25431/commits/05d1f2b5902124dcf7a087b4e3abf8e29fb8eb27)

Rugvip commented on beps/0005-definitive-plugin-listing/README.md:
````diff
@@ -0,0 +1,240 @@
+
+```yaml
+paths:
+  /.backstage/instanceInfo/features/installed:
````
So to be clear, just features in this instance right?

aramissennyeydd replied:
yes

Rugvip commented on beps/0005-definitive-plugin-listing/README.md:
```diff
@@ -0,0 +1,240 @@
+  };
+}
+
+interface DeploymentMetadataPluginExtensionPoint {
```
I think we'll want to provide these either with a custom implementation, or via configuration tbh, rather than mutating the service.

aramissennyeydd replied:
updated in [05d1f2b](https://github.com/backstage/backstage/pull/25431/commits/05d1f2b5902124dcf7a087b4e3abf8e29fb8eb27)

Rugvip commented on beps/0005-definitive-plugin-listing/README.md:
```diff
@@ -0,0 +1,240 @@
+
+### New `DeploymentMetadataPlugin`
+
+This will be a new core plugin that will be provided through a new package. The idea is that it will show all information across your entire deployment. One can imagine this hosting aggregate information around instances in the deployment, like their internal and external URLs.
```
What do you think about skipping the plugin for now and just making it a core service that consumes config? Then we could add a plugin if needed in the future

aramissennyeydd replied:
makes sense, done in [05d1f2b](https://github.com/backstage/backstage/pull/25431/commits/05d1f2b5902124dcf7a087b4e3abf8e29fb8eb27)

Rugvip reviewed (commented):
thank you for the updates! couple more things following up on the discussion in the SIG

aramissennyeydd committed:
address pr feedback

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd committed:
Update beps/0005-definitive-plugin-listing/README.md

Co-authored-by: Patrik Oldsberg <poldsberg@gmail.com>
Signed-off-by: Aramis Sennyey <159921952+aramissennyeydd@users.noreply.github.com>

Rugvip commented on beps/0005-definitive-plugin-listing/README.md:
```diff
@@ -166,21 +166,18 @@ interface BackstageInstance {
-interface DeploymentMetadataPluginExtensionPoint {
-  setInstances: (instances: BackstageInstance[]) => Promise<void>;
+interface DeploymentMetadataService {
+  listInstances: () => BackstageInstance[];
```
hmm, wonder if this should just be the same list of features rather than a list of instances. A slight issue I see with this is that we don't have the concept of routing to an instance, only plugins. It'd be an additional thing to add to the system.

aramissennyeydd replied:
Yeah, I call that out as a TODO below in the doc. I'll make that a bit more clear in the BEP - here's the original thread with Freben around plugins vs instances, https://github.com/backstage/backstage/pull/25431#discussion_r1722788174.

IMO we should stick with instances here and rely on the instance metadata service HTTP API to retrieve all metadata, the plugin-first approach is more discovery specific and in my mind is the custom discovery implementation using these 2 APIs. The big issue with just scraping the existing static config + list of the current instance's plugins is that it requires the config to be up to date with everything that's deployed. Couple of examples where I think that is a little rough,
1. Let's say you deploy the catalog + scaffolder to their own pods, but you have a single instance running smaller plugins, ie api-docs, techdocs, user-settings, etc. That single instance will have the correct list of plugins for the entire system, but unless you add each pluginId to your configs for catalog + scaffolder, they'd be inaccurate.
2. Similarly, if you want to dynamically add a plugin to a deployment, assuming you're just using the default discovery implementation, you'd need to adjust the config as well.

I think having the system metadata service provide instances and let the clients scrape instances themselves provides a bridge between the purely static and a more dynamic system that the future may hold.

Rugvip commented on beps/0005-definitive-plugin-listing/README.md:
```diff
@@ -166,21 +166,18 @@ interface BackstageInstance {
 
-interface DeploymentMetadataPluginExtensionPoint {
-  setInstances: (instances: BackstageInstance[]) => Promise<void>;
+interface DeploymentMetadataService {
```
perhaps SystemMetadataService or such? I'm worried deployment is a bit overloaded in kubernetes where it really just means a single instance in these terms

Rugvip commented on beps/0005-definitive-plugin-listing/README.md:
```diff
@@ -166,21 +166,18 @@ interface BackstageInstance {
+By default, this will
 a. be the exact same as `InstanceMetadataService` if we are not using `HostDiscovery` static config, or
-b. TODO: How will users define their endpoints statically such that we can access other instance's `InstanceMetadataService` HTTP APIs?
+b. use the `HostDiscovery` static config to fetch other instance's base URLs.
```
another option is to simply rely on local + explicitly configured plugins in discovery for this for now

Rugvip reviewed (commented):
:+1:,  couple more bits

aramissennyeydd committed:
update to system metadata service and mention a new config value

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd committed:
update system metadata service http api

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd commented:
@freben Was OOO, trying to get back on top of things - next steps here I believe is a review of https://github.com/backstage/backstage/pull/28726 from the maintainers side to drive a complete design for the BEP 🙂. That PR is pretty barebones so let me know if more fleshing out is required on my end before it can start to be reviewed.

freben reviewed (commented):
I was pondering this PR once again, reading it through. Some small notes, no great answers to the tricky questions

aramissennyeydd committed:
Apply suggestions from code review

Co-authored-by: Fredrik Adelöw <freben@gmail.com>
Signed-off-by: Aramis Sennyey <159921952+aramissennyeydd@users.noreply.github.com>

aramissennyeydd commented on beps/0005-definitive-plugin-listing/README.md:
```diff
@@ -0,0 +1,253 @@
+}
+
+interface SystemMetadataService {
+  getInstances: () => BackstageInstance[];
```
```suggestion
  getInstances: () => Promise<BackstageInstance[]>;
```

aramissennyeydd committed:
Update beps/0005-definitive-plugin-listing/README.md

Signed-off-by: Aramis Sennyey <159921952+aramissennyeydd@users.noreply.github.com>

Rugvip commented on beps/0005-definitive-plugin-listing/README.md:
```diff
@@ -0,0 +1,253 @@
+-->
+
+1. As a plugin builder, I can now get a list of currently installed features across my deployment.
+1. As a plugin builder, I can get a list of currently installed features for a single instance in my deployment.
```
I realized this is probably not something that a plugin builder should need to care about? Maybe it's interesting for an admin though?

Rugvip commented on beps/0005-definitive-plugin-listing/README.md:
```diff
@@ -0,0 +1,253 @@
+
+Split backends are a consistently difficult space to operate in and design for. There has been a growing desire for the framework to provide a way to get a list of the installed plugins. This was nearly impossible in the old backend, where plugins were hosted on denormalized routes and had non-standard startup sequences. In the new backend, this has become significantly more doable. Moving this f ...
+
+### Goals
```
Feeling that this area is perhaps lacking a little bit, there's probably something around building towards enabling dynamic discovery too?

Rugvip commented on beps/0005-definitive-plugin-listing/README.md:
```diff
@@ -0,0 +1,253 @@
+and make progress.
+-->
+
+1. The list of plugins will not change unless there is a config change, they will not be dynamically loaded.
```
From who's point of view is this? Either way it seems like features could be added or removed without a config change. Either in the local instance with a restart, or in a remote instance.

aramissennyeydd commented:
#28129 was recently merged - this BEP needs a round of updates to reflect conversation there.

aramissennyeydd committed:
reconcile BEP with shipped services and gateway plugin

Update the BEP to describe what actually shipped: RootInstanceMetadataService
(public) and RootSystemMetadataService (alpha) with plugin-centric shapes,
plus the @backstage/plugin-gateway-backend plugin for cross-instance routing.
Drop the unshipped HTTP API surface and rootUrl discovery config, move them
into Alternatives with rationale, and flip status to implemented.

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

aramissennyeydd committed:
remove notes and gateway plugin scope

Signed-off-by: aramissennyeydd <aramis.sennyey@doordash.com>

Rugvip reviewed (approved):
Should consider moving out of alpha, but separate from this really
